src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPools.cs

```diff
@@ -21,24 +21,55 @@ internal sealed class HttpConnectionPools : IDisposable
         /// <summary>The pools, indexed by endpoint.</summary>
         private readonly ConcurrentDictionary<HttpConnectionKey, HttpConnectionPool> _pools;
         /// <summary>Timer used to initiate cleaning of the pools.</summary>
-        private readonly Timer _cleaningTimer; // TODO #23149: Consider changing this to stop when _pools is empty.
+        private readonly Timer _cleaningTimer;
         /// <summary>The maximum number of connections allowed per pool. <see cref="int.MaxValue"/> indicates unlimited.</summary>
         private readonly int _maxConnectionsPerServer;
+        /// <summary>
+        /// Keeps track of whether or not the cleanup timer is running. It helps us avoid the expensive
+        /// <see cref="ConcurrentDictionary{TKey,TValue}.IsEmpty"/> call.
+        /// </summary>
+        private bool _timerIsRunning;
+        /// <summary>Object used to synchronize access to state in the pool.</summary>
+        private object SyncObj => _pools;
 
         /// <summary>Initializes the pools.</summary>
         /// <param name="maxConnectionsPerServer">The maximum number of connections allowed per pool. <see cref="int.MaxValue"/> indicates unlimited.</param>
         public HttpConnectionPools(int maxConnectionsPerServer)
         {
             _maxConnectionsPerServer = maxConnectionsPerServer;
             _pools = new ConcurrentDictionary<HttpConnectionKey, HttpConnectionPool>();
-            _cleaningTimer = new Timer(s => ((HttpConnectionPools)s).RemoveStalePools(), this, CleanPoolTimeoutMilliseconds, CleanPoolTimeoutMilliseconds);
+            // Start out with the timer not running, since we have no pools.
+            _cleaningTimer = new Timer(s => ((HttpConnectionPools)s).RemoveStalePools(), this, Timeout.Infinite, Timeout.Infinite);
+            _timerIsRunning = false;
```

Comment:
Nit: _timerIsRunning is initialized to its default value of false, so it doesn't actually need to be set to false.